packages/integration-aws-node/README.md

```diff
@@ -0,0 +1,105 @@
+# @backstage/integration-aws-node
+
+This package providers helpers for fetching AWS account credentials
+to be used by AWS SDK clients in backend packages and plugins.
+
+## Integrate new plugins
+
+Backend plugins can provide an AWS ARN or account ID to this library in order to
+retrieve a credentials provider for the relevant account that can be fed directly
+to an AWS SDK client.
+AWS SDK for Javascript V3 must be used.
+
+```typescript
+const awsCredentialsProvider = DefaultAwsCredentialsProvider.fromConfig(config);
+
+// provide the account ID explicitly
+const creds = await awsCredentialsProvider.getCredentials({ accountId });
```

Comment:
This I assume would generally only be useful for internal implementations right? Where you know your own accounts. Or would it make sense for plugins to have an additional layer of mapping different features to different account IDs?

Reply:
I would expect this to be part of a catalog annotation for the plugin if it needs an explicit account ID:
```yaml
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  annotations:
    # The account annotation for some fictional "AWS cost insights" plugin
    aws-cost-insights-account: '123456789012'
```

Or if relevant, part of the plugin's static configuration:
```yaml
# app-config.yaml
some-plugin:
  awsAccountId: '123456789012'
```

I can add some more detail for plugin authors around this to the documentation

Reply:
Got two separate thoughts on this, the higher level one being a concern whether this design really ends up fully addressing the need outlined in the original RFC? It certainly improves things either way, but I'm worried this will again push a lot of mapping to the plugins, where individual plugins need to figure out the `resource -> accountId` mapping. It takes me back to wondering whether there's a more generic way to be addressing the resources that one wants to access compared to using account IDs? That would likely be implemented as some form of mapping to account IDs, at which point your existing credentials logic takes over.

### Entity Query

Regarding the entity case, I feel that's a bit more straightforward, but it may be more powerful to instead have the interface be `.getCredentials({ entity })`. The credentials provider implementation that we ship by default might read it from an annotation, but you can override it with custom logic that might remove the need for populating the annotations in the first place.

You're hinting at specific credentials for different use-cases around an entity too though, which opens up a whole now level of fun. Scaling that out it seems like we'd be creating unique AWS credential annotations for every plugin that needs it, which then also needs to be populated in the catalog. Even with fallback annotations and other methods to slim things down that really seems like it would create a bit of a mess

### Step Back

I think perhaps it can be good to take a step back and see what's available too us, and if we're forced to ask users to fill in gaps somewhere. For example, considering the cost insights account for an entity, the question the plugin is asking is "what AWS account should I use to query cost insights data for this entity"? Let's assume that means access to the Cost Explorer API, I feel it's pretty clear what information is available to us, and could the implementation of that question then look like this? `.getCredentials({ service: 'ce', entity })`. Building on top of that you might ask for `.getCredentials({ service: 'eks' })` if you want EKS access that is not specific to any particular entity, or `.getCredentials({ service: 's3' })`. Combine that with the fact that the backend framework already provides you with the ID of the plugin that is asking for these credentials, and I feel you can get pretty far in mapping to desired accounts?

So I still think there are some gaps there though, what if you want access to a specific bucket? Would that be `.getCredentials({ service: 's3', bucket: 'techdocs' })`? That would get real messy, so perhaps you go for generic additional parameters then? Also a bit messy, the nicest case I feel is if there was a standard way to refer to these resources, for example `.getCredentials({ urn: 'eks:us-east-datapipelines' })` if you want to ask for access to a specific EKS cluster by name?  So I'm looking at what AWS provides there, and I'm wondering whether it makes sense to essentially take the ARN and extract the bits that plugins would reasonably know about, and make that the query? For example `.getCredentials({ service: 'eks', region: 'us-east-2', resourceType: 'cluster', resourceId: 'us-east-datapipelines' })`, or `.getCredentials({ service: 's3', resourceId: 'techdocs', entity })`. It's important that this would all map directly from the AWS data model, with no translation needed on our side.

With these types of queries in place we should be pretty free to implement different ways to do the `resource -> accountId` mapping in the credential provider implementation, and of course also allow integrators to roll their own.

Reply:
I think there are a few distinct use cases here:

**1. Plugins that act on specific AWS resources, like EKS clusters and DynamoDB tables.**

Examples: Proton plugin, Code suite plugin, CloudFormation plugin, Lambda plugin

A few background assumptions I'm making here, just to be explicit:
1. A plugin must be able to query AWS resources from many AWS accounts, because a Backstage user can own many AWS accounts.
1. An AWS resource must be queried using credentials for the account ID where the resource is located. There are various ways that AWS supports cross-account resource access, but they are not consistently used across services, so we will not rely on cross-account access in Backstage in order to standardize how plugins use credentials.
1. AWS resource names are not unique across accounts or regions (with the sole exception of S3 buckets). Two entities could refer to an EKS cluster with the same name 'my-cluster', but in reality refer to 2 different clusters in 2 different accounts.  Two entities could refer to an EKS cluster with the same name 'my-cluster' in the same account, but in reality refer to 2 different clusters in 2 different regions.
1. It is not possible to determine the account ID to use for querying a resource from the resource's name, region, service name, etc.  The user must explicitly provide the account ID to use in some form in order to select the correct credentials to use.
1. The unique identifier for AWS resources across accounts is the ARN, and (with the exception of S3 buckets) ARNs already have the account ID in them.

I am expecting that the resource ARN is typically the unique identifier used in static configuration and/or entity annotations to refer to an AWS resource that should be accessed by a plugin.  These plugins are not expected to handle account IDs; they are only expected to pass the ARN in its entirety.  But, they are then expected to parse the ARN to extract the resource name and region to construct whichever AWS client they need to use.

Example annotation:
```yaml
aws.amazon.com/aws-proton-service: arn:aws:proton:us-east-1:123456789012:service/my-proton-service
```

Example pseudo-code:
```typescript
import { parse } from '@aws-sdk/util-arn-parser'
...
    // Get the creds using the ARN from the entity annotation
   const creds = await this.awsCredentialsProvider.getCredentials({ arn: resourceArn });

   // Parse the ARN to get region and table name
   // Example ARN: arn:aws:dynamodb:us-east-2:123456789012:table/myDynamoDBTable
   const {region, resourceName} = parse(resourceArn);
   const tableName = resourceName.split("/")[1];

    // Query the table
    const client = new DynamoDbClient({
      region,
      credentialDefaultProvider: () => creds.provider,
    });
    const resp = await client
      .send(new DescribeTableCommand({ Table: tableName }));
```

However, this particular package is not opinionated about whether resource-level plugins choose to break apart the resource information into individual attributes, so they don't have to parse the ARN or if the resource name is not yet known (for example, when scaffolding/creating an AWS resource):

For example, region and account ID inputs for a fictional scaffolder action that generates a unique CloudFormation stack name and creates the stack:
```yaml
    - id: create-stack
      name: Create CloudFormation Stack
      action: aws:cloudformation:create-stack
     input:
        templatePath: ./template.yaml
        region: us-east-1
        accountId: 123456789012
```

Example pseudo-code:
```typescript
    // Get the creds using the account ID from the scaffolder action input
   const creds = await this.awsCredentialsProvider.getCredentials({ accountId });

    // Create the stack
    const client = new CloudFormationClient({
      region,
      credentialDefaultProvider: () => creds.provider,
    });
    const resp = await client
      .send(new CreateStackCommand({ ... }));
```

As another example, a fictional plugin that chooses to use multiple annotations for resource name, region, and account ID.
```yaml
aws.amazon.com/aws-cloudformation-stack-name: my-stack
aws.amazon.com/aws-cloudformation-stack-account: '123456789012'
aws.amazon.com/aws-cloudformation-stack-region: 'us-east-1'
```

Example pseudo-code:
```typescript
    // Get the creds using the account ID from the entity annotation
   const creds = await this.awsCredentialsProvider.getCredentials({ accountId });

    // Query the stack
    const client = new CloudFormationClient({
      region,
      credentialDefaultProvider: () => creds.provider,
    });
    const resp = await client
      .send(new DescribeStackCommand({ ... }));
```

**2. Plugins that act on S3 buckets.**

Examples: Techdocs plugin, S3 locations and discovery plugin

S3 buckets are weird in that their names are globally unique across accounts and the ARN does not contain an account ID, but they usually must be accessed using credentials from the account that owns the bucket.  There is no way from the bucket ARN or from any API call to determine what account owns a particular bucket, so some mapping of bucket name -> account credentials must be kept by the caller to access the bucket.

So in this case, static configuration and/or entity must have bucket name + account ID.  Example for techdocs plugin:
```
techdocs:
  publisher:
    type: 'awsS3'
    awsS3:
      bucketName: 'techdocs-storage'
      accountId: '123456789012'
```

The plugin uses the account ID to get the credentials, and the bucket name to access the bucket:
```ts
// get the creds for the account from the plugin configuration
const creds = await awsCredentialsProvider.getCredentials({ accountId });
// create S3 client with the creds
const client = new S3Client({
  region,
  credentialDefaultProvider: () => creds.provider,
});
// access the bucket
const resp = await client
  .send(new PutObjectCommand({Bucket: bucketName, ...
```

**3. Plugins that act on AWS accounts, like entity discovery and integration with account-level AWS services**

Examples: Any future plugins with Cost Explorer API, or resource discovery via EKS, AWS Organizations, Cloud Control API, etc

Since these plugins work at the account level, their configuration/entity annotations will typically contain the account ID to integrate with.

Example for a fictional EKS cluster locator:
```yaml
kubernetes:
  clusterLocatorMethods:
    - type: 'eks'
      accountId: '123456789012'
      region: 'us-east-1'
```

Example pseudo-code
```ts
// get the creds for the account from the plugin configuration
const creds = await awsCredentialsProvider.getCredentials({ accountId });

// List EKS clusters
const client = new EKSClient({
  region,
  credentialDefaultProvider: () => creds.provider,
});
const resp = await client
  .send(new ListClustersCommand({ });
```

Reply:
> It is not possible to determine the account ID to use for querying a resource from the resource's name, region, service name, etc. The user must explicitly provide the account ID to use in some form in order to select the correct credentials to use.

Yes, and the user in this case can be anything form an end user populating the catalog, to a Backstage integrator creating a custom AWS credentials provider.

I'll dial back the ambitions of flexibility here a bit though and let's simply evolve as needed. Other things that came to mind is how you'd handle something like separate accounts for prod/staging, or if you have large chunks of the org using the same accounts but perhaps slice them across different use-cases too. In the end we probably won't get it right from the beginning anyway so trying to cover more use-cases is likely to just bring us further in the wrong direction, so I'm happy to start small.

I'd like to still have at least one use of this new provider as part of this PR, what do you think about adding support for it in TechDocs?

Reply:
> I'd like to still have at least one use of this new provider as part of this PR, what do you think about adding support for it in TechDocs?

I actually had the same thought!  I just started working on upgrading TechDocs to the v3 Javascript SDK yesterday, since migrating over to this provider requires being on V3.  I'll open a separate PR for that change

Reply:
TechDocs support for this is now added